### ISSUE TITLE
How to add School Holidays?

### ISSUE THREAD

**Issue by afita** [ypid/opening_hours.js#126]
I tried to add School Holidays, following the example from DE, like so:

```
'SH': [
                {
                    name: 'Vacanța de iarnă',
                    2015: [ 12, 19, /* to */  1,  3 ]
                },
                {
                    name: 'Vacanţa intersemestrială',
                    2016: [  1, 30, /* to */  2,  7 ]
                },
                {
                    name: 'Vacanța de primăvară',
                    2016: [  4, 23, /* to */  5,  3 ]
                },
                {
                    name: 'Vacanța de vară',
                    2016: [  6, 18, /* to */  9,  4 ]
                }
            ]
```

Doesn't seem to work...

```
$ ./PH_SH_exporter.js --verbose --from=2015 --until=2017 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
$ # empty
```

Then I tried:

```
            'ro': {
                'SH': [
                    {
                        name: 'Vacanța de iarnă',
                        2015: [ 12, 19, /* to */  1,  3 ]
                    },
                    {
                        name: 'Vacanţa intersemestrială',
                        2016: [  1, 30, /* to */  2,  7 ]
                    },
                    {
                        name: 'Vacanța de primăvară',
                        2016: [  4, 23, /* to */  5,  3 ]
                    },
                    {
                        name: 'Vacanța de vară',
                        2016: [  6, 18, /* to */  9,  4 ]
                    }
                ]
            }
```

No joy:
```
$ ./PH_SH_exporter.js --verbose --from=2015 --until=2017 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please file a bug report here: https://github.com/ypid/opening_hours.js/issues?state=open. There are no holidays SH defined for country ro and state __state__. You can also add them: __repository_url__
Something went wrong. Please file issue at https://github.com/ypid/opening_hours.js/issues
```

**Comment by ypid** [comment]
For Germany every region/state has it’s own SH. And as Germany has currently the only supported SH, other defintions are not tested. Is that different for Romania?

**Comment by afita** [comment]
Yes, in Romania School Holidays are state wide. There's only a single set of School Holidays for the whole country; specified by law.

**Comment by ypid** [comment]
All right. Is probably only a bug in the SH selector function. I will look into it and use your definition when it works. Thanks for the contribution!

**Comment by ypid** [comment]
Yes, it was not tested and thus not working with SH country-wide. Fixed.

So, it your turn again :wink:. You might want to try to generate/scrape the SH definition somehow as I did it with the German SH definitions …

**Comment by afita** [comment]
Doesn't seem to work still...

I tried setting 'SH' country-wide (which makes sense from my point of view..). There is nowhere to scrape the School Holidays from, so I did them by hand following the ones from 'de'. Here's how the whole 'ro' looks like:

```
        'ro': { // {{{
            'PH': { // https://ro.wikipedia.org/wiki/S%C4%83rb%C4%83tori_publice_%C3%AEn_Rom%C3%A2nia#Zile_oficiale_de_s.C4.83rb.C4.83toare_.C3.AEn_care_nu_se_lucreaz.C4.83
                'Anul Nou'                                       : [  1,  1 ],
                'A doua zi de Anul Nou'                          : [  1,  2 ],
                'Ziua Unirii Principatelor Române (Ziua Unirii)' : [  1,  24 ],
                'Paștele ortodox'                                : [ 'orthodox easter',  0 ],
                'A doua zi de Paște ortodox'                     : [ 'orthodox easter',  1 ],
                'Ziua Muncii'                                    : [  5, 1 ],
                'Rusaliile'                                      : [ 'orthodox easter', 50 ],
                'A doua zi de Rusalii'                           : [ 'orthodox easter', 51 ],
                'Adormirea Maicii Domnului'                      : [  8, 15 ],
                'Sfântul Apostol Andrei'                         : [ 11, 30 ],
                'Ziua Națională (Ziua Marii Uniri)'              : [ 12,  1 ],
                'Crăciunul'                                      : [ 12, 25 ],
                'A doua zi de Crăciun'                           : [ 12, 26 ],
            },
            'SH': [
                {
                    name: 'Vacanța de iarnă',
                    2015: [ 12, 19, /* to */  1,  3 ]
                },
                {
                    name: 'Vacanţa intersemestrială',
                    2016: [  1, 30, /* to */  2,  7 ]
                },
                {
                    name: 'Vacanța de primăvară',
                    2016: [  4, 23, /* to */  5,  3 ]
                },
                {
                    name: 'Vacanța de vară',
                    2016: [  6, 18, /* to */  9,  4 ]
                }
            ]
        } // }}}
```

And here's the result:

```
$ ./PH_SH_exporter.js --verbose --from=2015 --to=2017 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5661
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition
```

Then I tried it this way:

```
        'ro': { // {{{
            'PH': { // https://ro.wikipedia.org/wiki/S%C4%83rb%C4%83tori_publice_%C3%AEn_Rom%C3%A2nia#Zile_oficiale_de_s.C4.83rb.C4.83toare_.C3.AEn_care_nu_se_lucreaz.C4.83
                'Anul Nou'                                       : [  1,  1 ],
                'A doua zi de Anul Nou'                          : [  1,  2 ],
                'Ziua Unirii Principatelor Române (Ziua Unirii)' : [  1,  24 ],
                'Paștele ortodox'                                : [ 'orthodox easter',  0 ],
                'A doua zi de Paște ortodox'                     : [ 'orthodox easter',  1 ],
                'Ziua Muncii'                                    : [  5, 1 ],
                'Rusaliile'                                      : [ 'orthodox easter', 50 ],
                'A doua zi de Rusalii'                           : [ 'orthodox easter', 51 ],
                'Adormirea Maicii Domnului'                      : [  8, 15 ],
                'Sfântul Apostol Andrei'                         : [ 11, 30 ],
                'Ziua Națională (Ziua Marii Uniri)'              : [ 12,  1 ],
                'Crăciunul'                                      : [ 12, 25 ],
                'A doua zi de Crăciun'                           : [ 12, 26 ],
            },
            'ro': {
                'SH': [
                    {
                        name: 'Vacanța de iarnă',
                        2015: [ 12, 19, /* to */  1,  3 ]
                    },
                    {
                        name: 'Vacanţa intersemestrială',
                        2016: [  1, 30, /* to */  2,  7 ]
                    },
                    {
                        name: 'Vacanța de primăvară',
                        2016: [  4, 23, /* to */  5,  3 ]
                    },
                    {
                        name: 'Vacanța de vară',
                        2016: [  6, 18, /* to */  9,  4 ]
                    }
                ]
            }
        } // }}}
```

Still not working:
```
        'ro': { // {{{
            'PH': { // https://ro.wikipedia.org/wiki/S%C4%83rb%C4%83tori_publice_%C3%AEn_Rom%C3%A2nia#Zile_oficiale_de_s.C4.83rb.C4.83toare_.C3.AEn_care_nu_se_lucreaz.C4.83
                'Anul Nou'                                       : [  1,  1 ],
                'A doua zi de Anul Nou'                          : [  1,  2 ],
                'Ziua Unirii Principatelor Române (Ziua Unirii)' : [  1,  24 ],
                'Paștele ortodox'                                : [ 'orthodox easter',  0 ],
                'A doua zi de Paște ortodox'                     : [ 'orthodox easter',  1 ],
                'Ziua Muncii'                                    : [  5, 1 ],
                'Rusaliile'                                      : [ 'orthodox easter', 50 ],
                'A doua zi de Rusalii'                           : [ 'orthodox easter', 51 ],
                'Adormirea Maicii Domnului'                      : [  8, 15 ],
                'Sfântul Apostol Andrei'                         : [ 11, 30 ],
                'Ziua Națională (Ziua Marii Uniri)'              : [ 12,  1 ],
                'Crăciunul'                                      : [ 12, 25 ],
                'A doua zi de Crăciun'                           : [ 12, 26 ],
            },
            'ro': {
                'SH': [
                    {
                        name: 'Vacanța de iarnă',
                        2015: [ 12, 19, /* to */  1,  3 ]
                    },
                    {
                        name: 'Vacanţa intersemestrială',
                        2016: [  1, 30, /* to */  2,  7 ]
                    },
                    {
                        name: 'Vacanța de primăvară',
                        2016: [  4, 23, /* to */  5,  3 ]
                    },
                    {
                        name: 'Vacanța de vară',
                        2016: [  6, 18, /* to */  9,  4 ]
                    }
                ]
            }
        } // }}}
```

**Comment by ypid** [comment]
Right, the error message is not very precise. I will improve this. The problem is that you tried to export the definition from 2015 until 2017 but only 2014, 2015 are defined.

Note that you should define multiple years like this.
```JavaScript
{
    name: 'Vacanța de iarnă',
    2014: [ 12, 19, /* to */  1,  3 ],
    2015: [ 12, 19, /* to */  1,  3 ],
},
```

PS: The definitions have to be online somewhere I would expect. I would prefer to have a script which can generate this list …

**Comment by ypid** [comment]
I find it easier to append each like like `2014: [ 12, 19, /* to */  1,  3 ]` by a comma, this makes it way easier to move blocks around. Is there a reason why you remove the comma on the last closing parentheses? Note the difference between JSON and objects in JavaScript.

**Comment by afita** [comment]
Still no... I added a new year and the comas. Now 'ro' looks like the following:

```
        'ro': { // {{{
            'PH': { // https://ro.wikipedia.org/wiki/S%C4%83rb%C4%83tori_publice_%C3%AEn_Rom%C3%A2nia#Zile_oficiale_de_s.C4.83rb.C4.83toare_.C3.AEn_care_nu_se_lucreaz.C4.83
                'Anul Nou'                                       : [  1,  1 ],
                'A doua zi de Anul Nou'                          : [  1,  2 ],
                'Ziua Unirii Principatelor Române (Ziua Unirii)' : [  1,  24 ],
                'Paștele ortodox'                                : [ 'orthodox easter',  0 ],
                'A doua zi de Paște ortodox'                     : [ 'orthodox easter',  1 ],
                'Ziua Muncii'                                    : [  5, 1 ],
                'Rusaliile'                                      : [ 'orthodox easter', 50 ],
                'A doua zi de Rusalii'                           : [ 'orthodox easter', 51 ],
                'Adormirea Maicii Domnului'                      : [  8, 15 ],
                'Sfântul Apostol Andrei'                         : [ 11, 30 ],
                'Ziua Națională (Ziua Marii Uniri)'              : [ 12,  1 ],
                'Crăciunul'                                      : [ 12, 25 ],
                'A doua zi de Crăciun'                           : [ 12, 26 ],
            },
            'SH': [
                {
                    name: 'Vacanța de iarnă',
                    2015: [ 12, 20, /* to */  1,  4 ],
                    2015: [ 12, 19, /* to */  1,  3 ],
                },
                {
                    name: 'Vacanţa intersemestrială',
                    2015: [  1, 31, /* to */  2,  8 ],
                    2016: [  1, 30, /* to */  2,  7 ],
                },
                {
                    name: 'Vacanța de primăvară',
                    2015: [  4, 11, /* to */  4, 19 ],
                    2016: [  4, 23, /* to */  5,  3 ],
                },
                {
                    name: 'Vacanța de vară',
                    2015: [  6, 20, /* to */  9, 13 ],
                    2016: [  6, 18, /* to */  9,  4 ],
                },
            ],
        }, // }}}
```

I do get something when I'm asking for --from=2015 --to=2015, but not with a range of years, like --2015 --2016 or older years:

```
$ ./PH_SH_exporter.js --verbose --from=2015 --to=2015 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
20151219-20151231 Vacanța de iarnă

$ ./PH_SH_exporter.js --verbose --from=2014 --to=2014 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2016 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2014 --to=2015 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2015 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition
```
In comparison with the 'de' country...

```
$ ./PH_SH_exporter.js --verbose --from=2015 --to=2015 --school-holidays --country de /tmp/ro_holidays.txt
20150101-20150105 Weihnachtsferien
20150330-20150410 Osterferien
20150526-20150606 Pfingstferien
20150730-20150912 Sommerferien
20151102-20151106 Herbstferien
20151223-20151231 Weihnachtsferien

$ ./PH_SH_exporter.js --verbose --from=2015 --to=2016 --school-holidays --country de /tmp/ro_holidays.txt
20150101-20150105 Weihnachtsferien
20150330-20150410 Osterferien
20150526-20150606 Pfingstferien
20150730-20150912 Sommerferien
20151102-20151106 Herbstferien
20151223-20160109 Weihnachtsferien
20160329-20160402 Osterferien
20160517-20160528 Pfingstferien
20160728-20160910 Sommerferien
20161102-20161104 Herbstferien
20161223-20161231 Weihnachtsferien
```

So something is clearly not working right...

**Comment by ypid** [comment]
Hm, right. Seems that I decided to not sort holiday definitions when it is not needed so holiday definition maintainers will need to keep that in mind. The following works:

```JavaScript
'SH': [
    {
        name: 'Vacanţa intersemestrială',
        2015: [  1, 31, /* to */  2,  8 ],
        2016: [  1, 30, /* to */  2,  7 ],
    },
    {
        name: 'Vacanța de primăvară',
        2015: [  4, 11, /* to */  4, 19 ],
        2016: [  4, 23, /* to */  5,  3 ],
    },
    {
        name: 'Vacanța de vară',
        2015: [  6, 20, /* to */  9, 13 ],
        2016: [  6, 18, /* to */  9,  4 ],
    },
    {
        name: 'Vacanța de iarnă',
        2015: [ 12, 20, /* to */  1,  4 ],
        2015: [ 12, 19, /* to */  1,  3 ],
    },
],
```

**Comment by afita** [comment]
Alright... It turns out I had a mistake in the 'Vacanța de iarnă' section: there were two entries with the same year, 2015. The first should have been 2014.

What is the criteria for the sorting? Alphabetical sorting based on 'name'? Chronological sorting based on the year, first date?

I sorted chronologically based on the first date. Now it looks like this:

```
'SH': [
                {
                    name: 'Vacanţa intersemestrială',
                    2015: [  1, 31, /* to */  2,  8 ],
                    2016: [  1, 30, /* to */  2,  7 ],
                },
                {
                    name: 'Vacanța de primăvară',
                    2015: [  4, 11, /* to */  4, 19 ],
                    2016: [  4, 23, /* to */  5,  3 ],
                },
                {
                    name: 'Vacanța de vară',
                    2015: [  6, 20, /* to */  9, 13 ],
                    2016: [  6, 18, /* to */  9,  4 ],
                },
                {
                    name: 'Vacanța de iarnă',
                    2014: [ 12, 20, /* to */  1,  4 ],
                    2015: [ 12, 19, /* to */  1,  3 ],
                },
            ],
```

And it almost works... for year 2015, not for 2014 & 2016:
```
$ ./PH_SH_exporter.js --verbose --from=2014 --to=2014 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2014 --to=2015 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2014 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2015 --to=2015 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
20150101-20150104 Vacanța de iarnă
20150131-20150208 Vacanţa intersemestrială
20150411-20150419 Vacanța de primăvară
20150620-20150913 Vacanța de vară
20151219-20151231 Vacanța de iarnă

$ ./PH_SH_exporter.js --verbose --from=2015 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

$ ./PH_SH_exporter.js --verbose --from=2016 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition

/home/adrian/work/FLOSS/opening_hours.js/opening_hours.js:5665
                        throw formatLibraryBugMessage(t('no SH definition', {
                              ^
An error occurred during evaluation of the value "SH". Please submit a pull request: https://github.com/ypid/opening_hours.js. no SH definition
```

**Comment by ypid** [comment]
The sorting looks good now. The errors you are getting are expected because the definition for 2014 and 2016 is not complete.

```JavaScript```
            'SH': [
                {
                    name: 'Vacanţa intersemestrială',
                    2015: [  1, 31, /* to */  2,  8 ],
                    2016: [  1, 30, /* to */  2,  7 ],
                },
                {
                    name: 'Vacanța de primăvară',
                    2015: [  4, 11, /* to */  4, 19 ],
                    2016: [  4, 23, /* to */  5,  3 ],
                },
                {
                    name: 'Vacanța de vară',
                    2015: [  6, 20, /* to */  9, 13 ],
                    2016: [  6, 18, /* to */  9,  4 ],
                },
                {
                    name: 'Vacanța de iarnă',
                    2014: [ 12, 20, /* to */  1,  4 ],
                    2015: [ 12, 19, /* to */  1,  3 ],
                    // 2016: [ 12, 19, 1,  3 ],],
                    // Comment out and 2016 will work.
                },
            ],
```

**Comment by afita** [comment]
Alright. Indeed it works if I add 2016 to 'Vacanța de iarnă':

```
$ ./PH_SH_exporter.js --verbose --from=2016 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
20160101-20160103 Vacanța de iarnă
20160130-20160207 Vacanţa intersemestrială
20160423-20160503 Vacanța de primăvară
20160618-20160904 Vacanța de vară
20161219-20161231 Vacanța de iarnă

$ ./PH_SH_exporter.js --verbose --from=2015 --to=2016 --school-holidays --country ro --region ro /tmp/ro_holidays.txt
20150101-20150104 Vacanța de iarnă
20150131-20150208 Vacanţa intersemestrială
20150411-20150419 Vacanța de primăvară
20150620-20150913 Vacanța de vară
20151219-20160103 Vacanța de iarnă
20160130-20160207 Vacanţa intersemestrială
20160423-20160503 Vacanța de primăvară
20160618-20160904 Vacanța de vară
20161219-20161231 Vacanța de iarnă
```
So the rules for the School Holidays are the following:
1. the entries should be manually sorted chronologically by the first date
2. a year should be present in each holiday entry in order to be taken into consideration. This is a bit odd and arbitrary because it's not known how the next school year will look like yet; it will be specified by a law; so for 2016 it's not known when the winter holiday - the first holiday from the 2016-2017school year - will be, but opening_hours.js requires it. It should be possible to have missing years from a school holiday entry.

**Comment by ypid** [comment]
> It should be possible to have missing years from a school holiday entry

That is something is I could try. The holiday code is one of the parts with the highest complexity in the lib so I avoid touching it unless it is absolutely required :wink:

Should I add your SH definition?

**Comment by afita** [comment]
I created pull request https://github.com/ypid/opening_hours.js/pull/129.

**Comment by ypid** [comment]
Thanks for your work :+1:

I wrote a test for the ro SH definitions. The SH code will throw a warning after 2016-09-05 so I would be happy if you could add the new definition before that date and users will not experience any problems. I did not check the holiday evaluation code in more deep again but one thing. I did not write a test in the code which throws the error on purpose when one year is not complete. And I think the current behavior makes sense because there is actually something missing so throwing an error is totally valid.

https://github.com/ypid/opening_hours.js/commit/0d7576dc2ad819ff4c82aadc0473187947b4693c#diff-1dd241c4cd3fd1dd89c570cee98b79ddR1176

**Comment by afita** [comment]
I don't think I understand what I need to do... Can you explain...?

**Comment by ypid** [comment]
Sorry, let me rephrase. As I understand you have done everything you could for now. PH/SH is defined. What I mean is that you are welcome to contribute more SH definitions when they are available. I would expect that the school holiday dates for 2016/2017 will be made public before 2016-09-05 so everything is good with the current way missing SH definitions are handled. The only thing I would want you to do is to add the SH definitions for 2016/2017 before 2016-09-05 so that users which use `SH` and test it for the current date with the evaluation tool will not get the error.

**Comment by afita** [comment]
Alright. I will monitor the Romanian law and update the School Holidays for 2016/2017 when the information will be available.

**Comment by ypid** [comment]
:+1:

**Comment by ypid** [comment]
Any updates yet? :wink:

**Comment by afita** [comment]
The School calendar for 2016-2017 has not been established yet. There is a ministerial project for public debate:

- http://edu.ro/index.php/pressrel/24558
- http://www.edu.ro/index.php/articles/24557
